### ISSUE TITLE
Compare Bidirectional Recurrence vs Unidirectional Recurrence

### ISSUE THREAD

**Issue by kdavis-mozilla** [mozilla/DeepSpeech#363]

**Comment by andi4191** [comment]
Dataset used: TED
Trainin time: 20:13:00
No. of Train epochs: 10
GPU instances used: 4

Results:

Unidirectional Recurrence Case:

I Test of Epoch 10 - WER: 0.448848, loss: 77.5554098553, mean edit distance: 0.209433
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.252548, mean edit distance: 0.076923
I - src: "no im serious"
I - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.650417, mean edit distance: 0.076923
I - src: "no im serious"
I - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.906622, mean edit distance: 0.076923
I - src: "no im serious"
I - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 3.964957, mean edit distance: 0.222222
I - src: "and i say"
I - res: "and i save"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 2.543439, mean edit distance: 0.100000
I - src: "farms that"
I - res: "forms that"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 3.561996, mean edit distance: 0.100000
I - src: "farms that"
I - res: "forms that"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 1.825124, mean edit distance: 0.333333
I - src: "and"
I - res: "nd"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.459296, mean edit distance: 0.500000
I - src: "so"
I - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.511882, mean edit distance: 0.500000
I - src: "so"
I - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 3.144284, mean edit distance: 1.000000
I - src: "so"
I - res: ""
I --------------------------------------------------------------------------------


Current Architecture - Bidirectional RNN with LSTM cells

Training time: 1:44:32

I FINISHED Optimization - training time: 1:44:32
I Test of Epoch 10 - WER: 0.415028, loss: 75.5818332566, mean edit distance: 0.196590
I --------------------------------------------------------------------------------
I WER: 0.166667, loss: 3.768760, mean edit distance: 0.035714
I  - src: "and treating them so i think"
I  - res: "and treating them so i thank"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 2.177231, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 2.362657, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.400000, loss: 3.249178, mean edit distance: 0.090909
I  - src: "its such a rich system"
I  - res: "it such a ric system"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 2.693307, mean edit distance: 0.090909
I  - src: "were lovers"
I  - res: "where lovers"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 1.175059, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 1.413706, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 1.456704, mean edit distance: 0.333333
I  - src: "and"
I  - res: "nd"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.545368, mean edit distance: 0.500000
I  - src: "okay"
I  - res: "ka"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.734726, mean edit distance: 0.333333
I  - src: "and"
I  - res: "nd"
I --------------------------------------------------------------------------------

**Comment by kdavis-mozilla** [comment]
When comparing _X_ against _Y_ you run both for 10 epochs.

How do you know that _X_ and _Y_ do not converge at different rates? In other words _X_ may take 1000 epochs to converge. _Y_ may take 100. But _X_ may have a lower WER when it converges. Thus at epoch 10 _X_ may look much worse than _Y_ as it's no where near convergence. However, _X_ may in fact be much better than _Y_ when it converges.

How do you know you have the optimal hyperparameters for both _X_ and _Y_? Did you check them?

**Comment by andi4191** [comment]
@kdavis-mozilla : Same set of hyperparameters were used on this case as well. In order to achieve our results on fair ground. I can think of following possible solutions:

1. Allowing early stopping for this on the same learning rate (Bug #667 PR in in progress) and increasing the number of epochs.
2. Tuning the learning rate for multiple benchmark variation algorithms by changing the learning rate for every run. It would require too much time to run as the running time for single epoch itself is like 2.5 hours.
3. We are using AdamOptimizer which itself is sort of combination of Momentum and AdaGrad optimizer. It uses the exponential decaying of learning rate. I am not sure how we come to optimal value of learning rate in our current model in the past.

Any suggestions?

**Comment by kdavis-mozilla** [comment]
@andi4191 Using the same set of hyperparameters is not a fair comparison. 

For example, it my be that the BRNN to achieve an optimal WER may require a learning rate that is smaller than the learning rate required by a RNN to achieve its optimal WER. What you want to be comparing is the optimal test set BRNN WER vs the optimal test set RNN WER.

What you want to be doing is tuning the hyperparameters, including the number of epochs, of the BRNN using the validation set, to optimize the validation set's WER. (For BRNN and the TED data set I've already done this.) Similarly, you want to tune the hyperparameters, including the number of epochs, of the RNN using the validation set to optimize the validation set's WER. Then you want to compare the optimal BRNN against the optimal RNN on the test set.

I think, due to computational time constraints, it would be fair to not worry about optimizing the WER but simply the loss, which is not so computationally heavy to obtain.

**Comment by andi4191** [comment]
Updating the results:
TED Dataset
UiRNN
Epochs: 15 (Optimal)
Learning rate : 0.0001
Inference time: 0:00:03

I Test of Epoch 15 - WER: 0.416078, loss: 81.7738748656, mean edit distance: 0.200658
I --------------------------------------------------------------------------------
I WER: 0.166667, loss: 2.806297, mean edit distance: 0.034483
I  - src: "and one of the most important"
I  - res: "an one of the most important"
I --------------------------------------------------------------------------------
I WER: 0.200000, loss: 2.080331, mean edit distance: 0.043478
I  - src: "it was a beautiful fish"
I  - res: "it was a beautiful dish"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 1.651304, mean edit distance: 0.043478
I  - src: "because this love story"
I  - res: "because this loves story"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 3.454031, mean edit distance: 0.125000
I  - src: "there is no feed"
I  - res: "there i is no feed"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.178140, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.624225, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 2.049138, mean edit distance: 0.111111
I  - src: "and i say"
I  - res: "and i saw"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 2.426721, mean edit distance: 0.111111
I  - src: "and i say"
I  - res: "and i saw"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 3.851700, mean edit distance: 0.142857
I  - src: "sixty thousand"
I  - res: "sixty thous"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.486991, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
